src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java

```diff
@@ -0,0 +1,185 @@
+package com.playmonumenta.networkrelay;
+
+import com.google.gson.JsonObject;
+import com.playmonumenta.networkrelay.util.MMLog;
+import java.util.Objects;
+import net.md_5.bungee.api.ProxyServer;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.api.event.PlayerDisconnectEvent;
+import net.md_5.bungee.api.event.PostLoginEvent;
+import net.md_5.bungee.api.plugin.Listener;
+import net.md_5.bungee.event.EventHandler;
+import net.md_5.bungee.event.EventPriority;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class RemotePlayerManagerBungee extends RemotePlayerManagerAbstraction implements Listener {
+	private static @MonotonicNonNull RemotePlayerManagerBungee INSTANCE = null;
+
+	private RemotePlayerManagerBungee() {
+		INSTANCE = this;
+		try {
+			for (String shard : NetworkRelayAPI.getOnlineShardNames()) {
+				MMLog.fine("Registering shard " + shard);
+				registerServer(shard);
+			}
+		} catch (Exception ex) {
+			MMLog.severe("Failed to get remote shard names");
+			throw new RuntimeException("Failed to get remote shard names", ex);
+		}
+
+		try {
+			NetworkRelayAPI.sendExpiringBroadcastMessage(REMOTE_PLAYER_REFRESH_CHANNEL,
+				new JsonObject(),
+				REMOTE_PLAYER_MESSAGE_TTL);
+		} catch (Exception ex) {
+			MMLog.severe("Failed to broadcast to channel " + REMOTE_PLAYER_REFRESH_CHANNEL);
+		}
+	}
+
+	protected static RemotePlayerManagerBungee getInstance() {
+		if (INSTANCE == null) {
+			INSTANCE = new RemotePlayerManagerBungee();
+		}
+		return INSTANCE;
+	}
+
+	protected static String getServerId() {
+		@Nullable String shardName = null;
+		try {
+			shardName = NetworkRelayAPI.getShardName();
+		} catch (Exception e) {
+			MMLog.severe("Failed to get shard name");
+		}
+		if (shardName == null) {
+			throw new RuntimeException("Got null shard name");
+		}
+		return shardName;
+	}
+
+	protected static RemotePlayerBungee fromLocal(ProxiedPlayer player, boolean isOnline) {
+		String targetShard = ""; // TODO: add a way to get the shard name. IT IS MUCH EASIER IN VELOCITY - usb
```

Comment:
It's better to make this nullable so it's more obvious the value is unknown/unset - unless you think there's a way to get this value from bungee before switching to velocity, in which case we should either implement it or switch to Velocity before this value gets used.

Reply:
The concern being that if we were to use this value as a tie-breaker when transferring players from one shard to another, we don't want it to claim they're on neither in an unintuitive way.